Improved the move_group_c++_interface tutorial

simonGoldstein opened the pull request:
### Description

The old Move Group C++ Interface Tutorial has some points for improvements. There were four things I wanted to improve:
1. The old tutorial didn't move the robot at all, and had the wrong command commented out to follow the plans it just made.
2. The old tutorial didn't move the object it attached to the robot.
3. The old tutorial had no functions or classes so it looked and acted like a script.
4. The old tutorial had not sub-tutorials so it left had many redundant and uninformative lines on the webpage.

I believe this solves the four problems I had with the old tutorial while also increasing the usefulness of the example code. To view what the new tutorial does the video below shows the full code being run.
https://www.youtube.com/watch?v=BcpNyNua9YM

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.trigger();
+    visual_tools.prompt("Press 'next' in the window to plan to the set Pose");
+
+    bool success = (move_group.plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
```
this should be in the sub-tutorial

simonGoldstein replied:
It kind of is. The strange comments a few lines up puts an identical line into the tutorial but I avoided the actual line because I didn't want the redundant visual_tools_ lines in the tutorial. I figured a strange comment a few lines up was better than a 1 line sub-tutorial.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+  void plan_constrained_path()
+  {
+    // Generating and setting the new Pose goal
+    target_pose.orientation.w = 0.0;
```
also, target_pose should not be a class variable. It should be created and passed around as necessary.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.prompt("Press 'next' to compare the constrained path with an unconstrained path");
+  }
+
+  void plan_unconstrained()
```
Making this into a class and putting the functionality into methods isn't that meaningful if the methods don't stand on their own. Set and clear path constraints should be their own methods and you should have a single method for `bool planToPose(const moveit_msgs::pose& target_pose, const moveit_msgs::Constraints& constraints, moveit::planning_interface::MoveGroupInterface::Plan &plan)` that returns success/failure as well as populating plan

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.prompt("Press 'next' in the window to move to the joint space goal");
+  }
+
+  void plan_pose_goal()
```
This should take in target_pose and return the plan as well as success/failure. Execution should be done separately. Make things modular and treat this more like a high level class for interacting with move_group.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    // END_SUB_TUTORIAL
+  }
+
+  void plan_joint_space_goal()
```
This should take in a joint_goal and return the plan as well as success/failure. Execution and visualization should be done in a separate class that get's re-used. Make things modular and treat this more like a high level class for interacting with move_group.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.prompt("Press 'next' in the window to move to the set Pose");
+  }
+
+  void plan_constrained_path()
```
see note on plan_unconstrained_path

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+  {
+  }
+
+  void setup_visualization()
```
methods should be camelCased.

simonGoldstein replied:
done

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.trigger();
+  }
+
+  void plan_cartesian_path()
```
This should take in a vector of target_goals and return the fraction of success. Execution and visualization should be done in a separate class that get's re-used. Make things modular and treat this more like a high level class for interacting with move_group.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.prompt("Press 'next' to move along the cartesian path");
+  }
+
+  void create_obstacle()
```
this should take in the size and pose of the object. As it stands, this doesn't demonstrate re-usable code. It is essentially a script broken up into clumps.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+
+    // RViz
+    ROS_INFO_NAMED("tutorial", "Add an object into the world");
+    visual_tools.deleteAllMarkers();
```
If you find yourself copy-pasting code a number of times, it is a good indication that you should put that functionality into a separate method. In this case,`publishStepAndWait(const std::string& step, const std::string& prompt_text);` would be one way to refactor this out.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    //
+    // Now, let's remove the collision object from the world.
+    planning_scene_interface.removeCollisionObjects(object_ids);
+    object_ids.pop_back();
```
This only removes the last value in the array but the above method removes the entire array.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.trigger();
+  }
+
+  void remove_obstacle()
```
This should take in object_ids and behave accordingly. Alternatively, it could be renamed removeAllCollisionObjects and simply remove everything

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.prompt("Press 'next' to create a new object");
+  }
+
+  void create_parcel()
```
What is a parcel? is it an obstacle? a box? Don't introduce new concepts unless you have to. addCollisionObject(id, pose, type) would be much better. Alternatively, you could create the collision_object elseware and pass it into this method.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    planning_scene_interface.addCollisionObjects(collision_objects);
+  }
+
+  void attach_parcel()
```
attach what parcel? Treat this as an api that someone else might want to use for their own use case and make these methods re-usable

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    // END_SUB_TUTORIAL
+  }
+
+  void plan_parcel()
```
what is plan_parcel? Isn't this just plan to pose? Before you add new methods, ask yourself, can this be accomplished using the existing api? Can I simply make another function more general to accomplish this same task? code duplication is bad practice. Reduce-reuse-recycle

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.trigger();
+  }
+
+  void detach_parcel()
```
pass in the id

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+    visual_tools.trigger();
+  }
+
+  void remove_parcel()
```
pass in the id. Check if the object is in the planning_scene first. Return true if it was found and removed

simonGoldstein replied:
done

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+
+    // Keeping the startstate updated with the robot is important. If the startstate isn't where
+    // the robot is then it will be unable to move.
+    move_group.setStartState(*move_group.getCurrentState());
```
Check the API. I believe you only need to do this to plan from a start state other than the current state. Otherwise you should just clear the start state and it will plan from the current state by default.

Also, this shouldn't be part of the execute command. execute should only execute. secondary effects like this are bad practice. If you want to set the start state, do it before planning. not after execution

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+
+    // END_SUB_TUTORIAL
+
+    visual_tools.prompt("Press 'next' in the RvizVisualToolsGui window to continue the demo");
```
Do you need this prompt here or should it be done outside of execute?

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+  {
+  }
+
+  void setup_visualization()
```
Also, every method should have a doxygen style comment explaining what it does, any input/output as well as any other information the user may want to know.

simonGoldstein commented on doc/move_group_interface/move_group_interface_tutorial.rst:
```diff
@@ -28,18 +27,21 @@ After a short moment, the RViz window should appear and look similar to the one
+ 4. The robot plans to the same pose goal whithout maintaining end-effector level.
+ 5. The robot moves its arm along the desired Cartesian path (a triangle down, right, up+left).
+ 6. A box object is added into the environment to the right of the arm.
     |B|
```
I don't really think that it needs timestamps on them all the video is only 46 seconds long

simonGoldstein commented on doc/move_group_interface/move_group_interface_tutorial.rst:
```diff
@@ -5,8 +5,7 @@ Move Group C++ Interface
 
-
-Watch this quick `YouTube video demo <https://youtu.be/_5siHkFQPBQ>`_ to see the power of the move group interface!
+Watch this quick `YouTube video demo <https://www.youtube.com/watch?v=BcpNyNua9YM>`_ to see the power of the move group interface!
```
The line is removed because of the embedded video

simonGoldstein commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+
+/* Author: Sachin Chitta, Dave Coleman, Mike Lautman, Simon Goldstein */
+
+// BEGIN_SUB_TUTORIAL includes
```
Done

simonGoldstein commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+  // MoveIt operates on sets of joints called "planning groups" and stores them in an
+  // object called the :joint_model_group:`JointModelGroup`. Throughout MoveIt the terms "planning group" and
+  // "joint model group" are used interchangably.
+  inline static const std::string PLANNING_GROUP = "panda_arm";
```
I updated this for all the class variables

BryceStevenWilley commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -1,394 +1,648 @@
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
```
This is a bad formatting change, should revert.

simonGoldstein replied:
This got changed when I clang formatted the document. I manually reverted it.

simonGoldstein commented:
I think I addressed all the feedback but I introduced a new error. When the object is attached it can't plan because the object is colliding with where it is attached. I will continue to search for a fix for that bug. It is possible there are now new things for me to fix in the meantime. Thanks again for the help.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  /**
+  * Publishes the given message above the robot
+  **/
+  void publishToRViz(std::string msg)
```
add an optional argument `bool blocking=false` and only call prompt if `blocking == true`

simonGoldstein replied:
I set it to default to true because I felt it resembled the average use of the function.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  /**
+  * Publishes the given message above the robot, and displays a path
+  **/
+  void publishToRViz(std::string msg, const moveit::planning_interface::MoveGroupInterface::Plan& plan)
```
No need for copy-pasta here. Call the other `publishToRviz` method or use a pointer for plan and set the default to `nullptr` so that you can optionally pass it in. Then in the code you can check 
```
if (plan) 
{
    visual_tools_.publishTrajectoryLine(plan.trajectory_, joint_model_group_);
}
```

mlautman replied:
see comment about blocking

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  /**
+  * Publishes the given message above the robot, displays a path, and points
+  **/
+  void publishToRViz(std::string msg, const moveit::planning_interface::MoveGroupInterface::Plan& plan, std::vector<geometry_msgs::Pose> points)
```
same

mlautman replied:
pass points vector by ref

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+
+    // We even can get a list of all the groups in the robot.
+    ROS_INFO_NAMED(name_, "Available Planning Groups:");
+    std::copy(move_group_.getJointModelGroupNames().begin(), move_group_.getJointModelGroupNames().end(),
```
Don't print to cout if you can use the ros console functionality. 
Better to use string stream and pass that to `ROS_INFO_NAMED(name_, ss.toString())` <--- likely incorrect syntax

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  * Populates the plan with a path to the given joint space goal
+  * Returns the success of planning
+  **/
+  bool planJointSpaceGoal( const std::vector<double> joint_group_positions, moveit::planning_interface::MoveGroupInterface::Plan& plan)
```
much better! pass `joint_group_positions` in by reference to avoid data copy. 

(`const std::vector<double> &joint_group_positions`)

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  * Populates the plan with a path to the pose goal
+  * Returns the success of planning
+  **/
+  bool planPoseGoal(geometry_msgs::Pose pose, moveit::planning_interface::MoveGroupInterface::Plan& plan, bool resetStartState=false)
```
```suggestion
  bool planPoseGoal(const geometry_msgs::Pose& pose, moveit::planning_interface::MoveGroupInterface::Plan& plan, bool resetStartState=false)
```

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  {
+    // Someitmes it can be useful to plan a path that doesn't start where the current robot is. If that is done it is important
+    // to reset the planning start state back to where the robot is when done. This is also used when an object is attached.
+    if (resetStartState)
```
Probably minor, but the default behavior should be to use the current state. 
Better would be to optionally pass in a start state and if it is not null then setStartState. In this approach you will have to clear the start state at the end.

simonGoldstein replied:
When you say in this approach, do you mean in the current approach or in your proposed new method?

mlautman replied:
proposed new approach

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  /**
+  * Adds the given planning constraints
+  **/
+  void addConstraint(moveit_msgs::Constraints test_constraints)
```
```suggestion
  void addConstraint(const moveit_msgs::Constraints& test_constraints)
```

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+    // Planning with constraints can be slow because every sample must call an inverse kinematics
+    // solver. Lets increase the planning time from the default 5 seconds to be sure the planner
+    // has enough time to succeed.
+    move_group_.setPlanningTime(10.0);
```
This does not belong here. It is a secondary affect that does not reflect the method name and is undocumented. If you want to setPlanningTime do it in a more appropriate place.

simonGoldstein replied:
It is now a new unique function

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+    // BEGIN_SUB_TUTORIAL remove_constraint
+    // After moving it is often useful to revert back to the default settings.
+    move_group_.clearPathConstraints();
+    move_group_.setPlanningTime(5.0);
```
same

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  /**
+  * Populates the plan with a cartesian path following the given points in order
+  **/
+  double planCartesianPath(std::vector<geometry_msgs::Pose> target_poses, moveit::planning_interface::MoveGroupInterface::Plan& plan)
```
Pass by ref for complex data types and use `const` whenever you aren't changing that input

```suggestion
  double planCartesianPath(const std::vector<geometry_msgs::Pose>& target_poses, moveit::planning_interface::MoveGroupInterface::Plan& plan)
```

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+    //
+    // Cartesian motions are frequently needed to be slower for actions such as approach and
+    // retreat grasp motions. Here we demonstrate how to reduce the speed of the robot arm via
+    // a scaling factor of the maxiumum speed of each joint. Note this is not the speed of the
```
"Note this is the joint velocities not the end effector velocities."

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+    // Moving the robot with this plan is the same as before.
+    //
+    // When finished remember to reset the scalingfactor to return to normal speeds.
+    move_group_.setMaxVelocityScalingFactor(1);
```
nit: If the user set this elseware, this could trip up their code. I would maybe copy the current factor to a local variable and then use it here.

simonGoldstein replied:
MoveGroup only has a setter for this and no getter

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  * geometry_msgs::Pose box_pose - The location
+  * std::string name - The object ID
+  **/
+  void createObject(shape_msgs::SolidPrimitive primitive, geometry_msgs::Pose box_pose, std::string name)
```
use const and references to avoid data copy

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
+  // Create the plan we will be using
+  moveit::planning_interface::MoveGroupInterface::Plan plan;
+  // This is a vector of locations for moveing to multiple targets.
+  std::vector<geometry_msgs::Pose> target_poses;
```
initialize this closer to where you will use it.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,481 @@
-  // We can also print the name of the end-effector link for this group.
-  ROS_INFO_NAMED("tutorial", "End effector link: %s", move_group.getEndEffectorLink().c_str());
+  // Create the plan we will be using
+  moveit::planning_interface::MoveGroupInterface::Plan plan;
```
move this to where you will use it

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -78,74 +78,38 @@ class MoveGroupInterfaceTutorial
   * Publishes the given message above the robot, displays a path, and points
   **/
-  void publishToRViz(std::string msg, const moveit::planning_interface::MoveGroupInterface::Plan& plan, std::vector<geometry_msgs::Pose> points)
+  void publishToRViz(std::string msg, const moveit::planning_interface::MoveGroupInterface::Plan* plan=nullptr, const std::vector<geometry_msgs::Pose>* points=nullptr, bool blocking=true, bool clear_prev=true)
```
In C++ we don't use raw pointers unless we have to. If you need a pointer you should use a smart pointer like `std::shared_ptr` let me know if you need examples to work from

mlautman replied:
Also, if you're only using `plan->trajectory_` then pass that in directly rather than passing in plan.


eg 
```
void publishToRViz(std::string msg, const moveit_msgs::RobotTrajectory& plan = moveit_msgs::RobotTrajectory())
{
  ...
  if (plan.joint_trajectory.points.size())
  {
    ...
  }
}
```

mlautman replied:
Rather than use a pointer to a std::vector, you should make the default a std::vector<geometry_msgs::Pose> of length 0

simonGoldstein replied:
Done

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -78,74 +78,38 @@ class MoveGroupInterfaceTutorial
-    for( geometry_msgs::Pose point : points )
-      visual_tools_.publishAxisLabeled(point, "pose");
+    if (points)
+      for( geometry_msgs::Pose point : *points )
```
Assuming you remove the raw pointers, this should be:

```
for(const geometry_msgs::Pose &point : points )
￼
```

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -78,74 +78,38 @@ class MoveGroupInterfaceTutorial
+  void publishToRViz(std::string msg, const moveit::planning_interface::MoveGroupInterface::Plan* plan=nullptr, const std::vector<geometry_msgs::Pose>* points=nullptr, bool blocking=true, bool clear_prev=true)
   {
-    // BEGIN_SUB_TUTORIAL visualization
+    // BEGIN_SUB_TUTORIAL visualization2
```
if we no longer have a `visualization` we shouldn't have a `visualizatoin2`

simonGoldstein replied:
No we do not

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -441,13 +414,9 @@ int main(int argc, char** argv)
-
   // Plan and move to Joint space goal
-  move_group_interface_tutorial.publishToRViz("Joint Space Goal");
+  std::string publish_message = "Joint Space Goal";
```
you could simply pass this in directly to publishToRviz rather than use this intermediate var

simonGoldstein replied:
If I am publishing the same message to RViz 3-4 times I thought it made it easier to have an intermediate variable.

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -239,22 +202,33 @@ class MoveGroupInterfaceTutorial
   }
 
+  /**
+   *
```
This feels rather empty.. remove it

simonGoldstein replied:
Added a small description

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -456,11 +425,16 @@ int main(int argc, char** argv)
   move_group_interface_tutorial.move(plan);
 
-  move_group_interface_tutorial.publishToRViz("Pose Goal");
+  publish_message = "Pose Goal";
```
let's keep minor changes like this to a minimum and focus on getting this merged

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -95,19 +94,18 @@ class MoveGroupInterfaceTutorial
     // To publish a planed path.
-    if (plan)
-      visual_tools_.publishTrajectoryLine(plan->trajectory_, joint_model_group_);
+    if (trajectory.joint_trajectory.header.frame_id != "")
```
nit: you can check if a string is empty

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -95,19 +94,18 @@ class MoveGroupInterfaceTutorial
     // To publish points.
-    if (points)
-      for( geometry_msgs::Pose point : *points )
+    if (points.size() != 0)
```
nit: no need to check this

mlautman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -164,10 +163,13 @@ class MoveGroupInterfaceTutorial
-    // Someitmes it can be useful to plan a path that doesn't start where the current robot is. If that is done it is important
-    // to reset the planning start state back to where the robot is when done. This is also used when an object is attached.
+    // Someitmes it can be useful to plan a path that doesn't start where the current robot is. If that is done it is
+    // important
```
fix comments

davetcoleman commented on doc/move_group_interface/move_group_interface_tutorial.rst:
```diff
@@ -28,20 +29,19 @@ After a short moment, the RViz window should appear and look similar to the one
- 8. The object is detached from the wrist (its color will change back to green).
- 9. The object is removed from the environment.
-
-.. |B| image:: ./move_group_interface_tutorial_robot_with_box.png
```
why remove this image?

simonGoldstein replied:
The image was from the old tutorial and I removed it form the middle of the bulleted list because I found it intrusive. I decided not to replace it above or bellow because there is already an embedded video above.

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  **/
+  MoveGroupInterfaceTutorial()
+  {
+  }
```
There is no need for this constructor as its empty

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  bool planPoseGoal(const geometry_msgs::Pose& pose, moveit::planning_interface::MoveGroupInterface::Plan& plan,
+                    bool resetStartState = false)
+  {
+    // Someitmes it can be useful to plan a path that doesn't start where the current robot is. If that is done it is
```
spelling

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  /**
+  * Adds the given planning constraints
+  **/
+  void addConstraint(const moveit_msgs::Constraints& test_constraints)
```
why are we wrapping single function calls with a function? this function provides no value other than adding complexity to what should be a simple tutorial

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  /**
+  * Removes planning constraints
+  **/
+  void removeConstraints()
```
why are we wrapping single function calls with a function? this function provides no value other than adding complexity to what should be a simple tutorial

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  /**
+   * This sets the planning time to a number of seconds.
+   **/
+  void setPlanningTime(double time = 5.0)
```
why are we wrapping single function calls with a function? this function provides no value other than adding complexity to what should be a simple tutorial

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  * geometry_msgs::Pose box_pose - The location
+  * std::string name - The object ID
+  **/
+  void createObject(const shape_msgs::SolidPrimitive& primitive, const geometry_msgs::Pose& box_pose, std::string name)
```
function needs more specific name... it only creates boxes, right?

simonGoldstein replied:
The shape_msgs::SolidPrimitive can be a cone, sphere, square or cylinder. It is only used to create boxes but it could be used for other shaped object creation.

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  * Attaches a object to the robot
+  * std::string id - ID of the object to attach
+  **/
+  void attachObject(std::string id)
```
why are we wrapping single function calls with a function? this function provides no value other than adding complexity to what should be a simple tutorial

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
+  /**
+  * Moves the robot according to the given plan
+  **/
+  void move(const moveit::planning_interface::MoveGroupInterface::Plan& plan)
```
why are we wrapping single function calls with a function? this function provides no value other than adding complexity to what should be a simple tutorial

davetcoleman commented on doc/move_group_interface/src/move_group_interface_tutorial.cpp:
```diff
@@ -32,169 +32,456 @@
-  // We can print the name of the reference frame for this robot.
-  ROS_INFO_NAMED("tutorial", "Planning frame: %s", move_group.getPlanningFrame().c_str());
+  // Create and setup the MoveGroupInterfaceTutuorial instance
+  MoveGroupInterfaceTutorial move_group_interface_tutorial;
```
I don't like this class. Its basically an interface for an interface to MoveGroup, which is a wrapper to MoveIt already... this only further complicates using MoveIt.

davetcoleman commented:
> The old tutorial had no functions or classes so it looked and acted like a script.

IMHO, tutorials are often best as a script, not as a full application. I know this is an unfortunate opinion given the work put into this, but I have put a lot of [cleanup effort into this tutorial myself](https://github.com/ros-planning/moveit_tutorials/pull/42) and agree with @BryceStevenWilley

simonGoldstein commented:
@mlautman requested this

davetcoleman commented:
if I actually saw tests (in this or the python tutorial) this change would be a bit more motivating...

davetcoleman commented:
@BryceStevenWilley and I are not in favor of this OR, and @simonGoldstein's internship ended. Closing.
